### ISSUE TITLE
how to delete node from old graph？

### ISSUE THREAD

**Issue by neverPick** [tensorflow/tensorflow#2245]
I build retrain_image for 5 new flowers,and got a pb file ,87M,has 1001 nodes.I put it in android demo.Every run costs 2S.It is terrorable.I guess,because of too many unusefull nodes.So,how to delete nodes those I wish in retraining?

**Comment by zheng-xq** [comment]
One way is to change the default graph, and not putting all the nodes on the same graph. 

with tf.Graph().as_default():

Adding @mrry to point out if there are better ways.

**Comment by mrry** [comment]
If a graph node is unused, it will not be executed, and hence it won't contribute to the step time (apart from possibly the first step, when the graph is first built and pruned). There's no API for deleting nodes in Python, but you can try using the [`extract_sub_graph()`](https://github.com/tensorflow/tensorflow/blob/5f4524928af5ba5727fa97a35ddd5da72be0c476/tensorflow/python/client/graph_util.py#L125) function to prune the graph to only the nodes that are needed. (Note that doing this is unlikely to improve performance though.)

**Comment by Arsakes** [comment]
`extract_sub_graph()` is hard to use.
A **variable** consist of few nodes and there is no single node that has an outgoing connection to each other. 
_Thanks to this fact `export_sub_graph()` breaks variable structure and effectivly removes variable from graph. 
Is there a smart way around it? (I didn't try explicitely listing all the nodes for each variable used by source graph - but this wouldn't be super efficient and clean).

**Comment by yaroslavvb** [comment]
There's a more advanced subgraph extractor proposed in https://github.com/tensorflow/tensorflow/pull/5802

**Comment by drcrook1** [comment]
I think I need this feature.  I trained a graph with queue runners and now need to run inference against it with feed_dict.  When I load the meta graph the first part of the graph is the queue runner input pipeline; so when I call predict; it hangs unless I start the queue runners; when I start the queue runners they are empty because I'm not putting anything on the queue; I have no idea how to enqueue items...  The inference images never hit disk until the whole process is done and even then only in collection mode.

**Comment by mrry** [comment]
@drcrook1 This sounds like a problem for the [TensorFlow graph transform tool](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md). It enables you to specify an input node (which in your case would probably come after the `"DequeueMany"` op from the queue), and will prune the irrelevant nodes for you.

/cc @petewarden

**Comment by ljstrnadiii** [comment]
@mrry any plans to add a feature to remove dropout in the tensorflow graph transform tool? I am running into the issue that nvidia's tensorrt does not support the op, which forces me to manually remove the nodes. Or is there an easy way to do that already with optimizing for deployment?

**Comment by mrry** [comment]
@ljstrnadiii I'm not sure what the plans are for the graph transform tool, but @petewarden would know. It might be worth creating a new issue with this high-level feature request.

**Comment by ljstrnadiii** [comment]
Thanks @mrry .

**Comment by ysw361564483** [comment]
@drcrook1 I have same problem, have you fixed it ?

**Comment by ysw361564483** [comment]
@mrry  I have seen [TensorFlow graph transform tool](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md#merge_duplicate_nodes), but I don't figure out which transform operation can implement "specify an input node"...

**Comment by jc3000** [comment]
I have the same issue as @ysw361564483. Anyone knows how to set an input (e.g. phase_train) which is placeholder with graph transform tool for inference optimization?

**Comment by NPetsky** [comment]
@drcrook1  I have added manually an input placeholder node with right dimensions and deleted the whole queue related nodes in the pbtxt file, but for this you have to know where to connect your new node (looking at the graph in tensorboard helped me)

**Comment by zhongzhaoji** [comment]
> @mrry any plans to add a feature to remove dropout in the tensorflow graph transform tool? I am running into the issue that nvidia's tensorrt does not support the op, which forces me to manually remove the nodes. Or is there an easy way to do that already with optimizing for deployment?

I have the same problem that need to remove unspported op in using tensorrt,do you have some way to solve this problem?Thanks you
